Add recommendation for ZkSystemSpec parameters (#95)
google/longfellow-zk: docs/content/en/docs/zk-system-spec.md

---
title: "Longfellow ZK System spec parameters"
linkTitle: "ZkSystemSpec"
weight: 100
description: >-
     List of ZK System spec parameters expected by Longfellow ZK library
---

ISO 18013-5 and and OpenId4VP allow RPs to request ZKP responses and specify any ZKP system and some parameters required to make sure RP will be able to verify the proof correctly.

> The list of the parameters and their values are not defined in the standard and specific for each ZKP system. 

Reviewer: ```suggestion
The list of the parameters and their values are not defined in the standard and are specific for each ZKP system. 
```

> This document describes which paramters should be used by users of Longfellow ZK System. Those paramters are currently used by Google Wallet and Multipaz Wallet/RP.

Reviewer: ```suggestion
This document describes which parameters should be used by users of Longfellow ZK System. Those parameters are currently used by Google Wallet and Multipaz Wallet/RP.
```

For the Longfellow ZK the following parameters and values should be used:
* **zkSystemId**: <any value, we recommend using circuit_hash to avoid confusion>
* **system**: 'longfellow-libzk-v1'

The following parameters must be present:
* **circuit_hash**: Hash of the used circuit
* **num_attributes**: number of requested attributes
* **version**: circuit version
* **block_enc_hash**: The block_enc parameter for the ZK proof of the hash component

> * **block_enc_sig**: block_enc parameter for the ZK proof of the signature component

Reviewer: ```suggestion
* **circuit_hash**: The hash of the used circuit.
* **num_attributes**: The number of requested attributes.
* **version**: The circuit version.
* **block_enc_hash**: The `block_enc` parameter for the ZK proof of the hash component.
* **block_enc_sig**: The `block_enc` parameter for the ZK proof of the signature component.
```

> **The paramters change relatively often, we are releasing new cirucits regularly, so it's strongly recommended not to build ZKSpecs manually or hardcode them, but instead use [kZkSpecs](https://github.com/google/longfellow-zk/blob/main/lib/circuits/mdoc/zk_spec.cc) as a source of truth for the paramters

Reviewer: ```suggestion
**The paramters change relatively often. We are releasing new circuits regularly, so it's strongly recommended not to build ZKSpecs manually nor hardcode them, but instead to use [`kZkSpecs`](https://github.com/google/longfellow-zk/blob/main/lib/circuits/mdoc/zk_spec.cc) as a source of truth for the parameters
```

> and convert it to CBOR/JSON as needed for corresponsing protocols.**

Reviewer: ```suggestion
and convert it to CBOR/JSON as needed for corresponding protocols.**
```

## ZK Spec Parameters in ISO 18013-5

> [ISO 18013-5 Second Edition](https://github.com/ISOWG10/ISO-18013/tree/main/Working%20Documents) section 10.2.7 defines ZkSystemSpec like this:

Reviewer: ```suggestion
[ISO 18013-5 Second Edition](https://github.com/ISOWG10/ISO-18013/tree/main/Working%20Documents) section 10.2.7 defines `ZkSystemSpec` like this:
```
